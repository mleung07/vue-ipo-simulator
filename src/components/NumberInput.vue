<template>
  <!-- TODO: make the input field invisible & display formatted value at bottom -->
  <input type="text" :value="displayValue" @input="inputHandler" required />
</template>

<script>
export default {
  name: "NumberInput",
  props: ["value"],
  data: function() {
    return {
      displayValue: "",
    };
  },
  // computed: {
  //   displayValue: {
  //     get: function() {
  //       if (!this.value) {
  //         return "";
  //       }

  //       return new Intl.NumberFormat("en-HK").format(this.value);
  //     },
  //     set: function(modifiedValue) {
  //       const newValue = parseFloat(modifiedValue.replace(/[^\d.]/g, ""));

  //       if (isNaN(newValue)) {
  //         return;
  //       }

  //       this.$emit("input", newValue);
  //     },
  //   },
  // },
  // watch: {
  //   displayValue: function(newValue, oldValue) {
  //     console.log(oldValue, newValue);
  //     const rule = /[^0-9.,]/;
  //     if (rule.test(newValue)) {
  //       console.log("setting to prev value");
  //       this.displayValue = oldValue;
  //       this.$emit("input", oldValue);
  //     }
  //   },
  // },
  methods: {
    inputHandler(ev) {
      const value = ev.target.value;
      const number = parseFloat(value.replace(/[^\d.]/g, ""));
      this.$emit("input", number);
      if (!isNaN(number)) {
        this.displayValue = new Intl.NumberFormat("en-HK").format(number);
      }
      // this.$forceUpdate();
    },
    convertToCurrency(number) {
      return number;
    },
  },
};
</script>

<style lang="scss" scoped>
input {
  font-size: 16px;
  padding: 8px 0px;
  width: 100%;
  border: none;
  border-bottom: 1px solid #aaa;
}
input:focus {
  outline: none;
}
</style>
