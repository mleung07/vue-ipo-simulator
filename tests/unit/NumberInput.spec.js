import { mount, shallowMount } from "@vue/test-utils";
import NumberInput from "@/components/NumberInput.vue";

describe("NumberInput.vue", () => {
  it("matches snapshot", () => {
    const wrapper = mount(NumberInput);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("display correctly after input 1", () => {
    const wrapper = shallowMount(NumberInput);
    const input = wrapper.find("input");
    input.element.value = 1;
    input.trigger("input");
    expect(wrapper.vm.displayValue).toBe("1");
    expect(wrapper.emitted().input).toBeTruthy();
  });

  it("display correctly after input 1m", () => {
    const wrapper = shallowMount(NumberInput);
    const input = wrapper.find("input");
    input.element.value = 1000000;
    input.trigger("input");
    expect(wrapper.vm.displayValue).toBe("1,000,000");
    expect(wrapper.emitted().input).toBeTruthy();
  });

  it("display correctly after input NaN", () => {
    const wrapper = shallowMount(NumberInput);
    const input = wrapper.find("input");
    input.element.value = "a";
    input.trigger("input");
    expect(wrapper.vm.displayValue).toBe("");
    expect(wrapper.emitted().input[0]).toBeTruthy();
  });
});
