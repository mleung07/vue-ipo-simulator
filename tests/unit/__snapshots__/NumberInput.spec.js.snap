// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput.vue matches snapshot 1`] = `
<input
  required="required"
  type="text"
/>
`;
